src/rgw/driver/sfs/sqlite/sqlite_versioned_objects.cc

```diff
@@ -13,110 +13,380 @@
+    if (last_version) {
+      ret_value = *last_version;
+    }
+  }
+  return ret_value;
+}
+
+std::optional<DBVersionedObject>
+SQLiteVersionedObjects::delete_version_and_get_previous_transact(uint id) {
+  try {
+    auto storage = conn->get_storage();
+    auto transaction = storage.transaction_guard();
+    auto version = storage.get_pointer<DBVersionedObject>(id);
+    std::optional<DBVersionedObject> ret_value;
+    if (version != nullptr) {
+      auto object_id = version->object_id;
+      storage.remove<DBVersionedObject>(id);
+      // get the last version of the object now
+      auto max_commit_time_ids = storage.select(
+          columns(&DBVersionedObject::id, max(&DBVersionedObject::commit_time)),
+          where(
+              is_equal(&DBVersionedObject::object_id, object_id) and
+              is_not_equal(
+                  &DBVersionedObject::object_state, ObjectState::DELETED
+              )
+          ),
+          group_by(&DBVersionedObject::id),
+          order_by(&DBVersionedObject::id).desc()
+      );
+      auto found_value = max_commit_time_ids.size() &&
+                         std::get<1>(max_commit_time_ids[0]) != nullptr;
+      if (found_value) {
+        // if not value is found could be, for example, if lifecycle deleted all
+        // non current versions before.
+        auto last_version_id = std::get<0>(max_commit_time_ids[0]);
+        auto last_version =
+            storage.get_pointer<DBVersionedObject>(last_version_id);
+        if (last_version) {
+          ret_value = *last_version;
+        }
+      }
+      transaction.commit();
+    }
+    return ret_value;
+  } catch (const std::system_error& e) {
+    // throw exception (will be caught later in the sfs logic)
+    // TODO revisit this when error handling is defined
+    throw(e);
+  }
+}
+
+std::optional<DBVersionedObject> SQLiteVersionedObjects::_get_versioned_object(
+    const std::string& bucket_id, const std::string& object_name,
+    const std::string& version_id
 ) const {
   auto storage = conn->get_storage();
-  auto last_version_id = storage.max(
+  auto ids = storage.select(
       &DBVersionedObject::id,
-      where(c(&DBVersionedObject::object_id) = object_id.to_string())
+      inner_join<DBObject>(
+          on(is_equal(&DBObject::uuid, &DBVersionedObject::object_id))
+      ),
+      where(
+          is_not_equal(
+              &DBVersionedObject::object_state, ObjectState::DELETED
+          ) and
+          is_equal(&DBObject::bucket_id, bucket_id) and
+          is_equal(&DBObject::name, object_name) and
+          is_equal(&DBVersionedObject::version_id, version_id)
+      )
   );
-  std::optional<DBOPVersionedObjectInfo> ret_value;
-  if (last_version_id) {
-    auto last_version =
-        storage.get_pointer<DBVersionedObject>(*last_version_id);
+  // TODO return an error if this returns more than 1 version?
+  // Only 1 object with no deleted versions should be present
```

Comment:
I can do that, yeah... That's some of the things we'd need to discuss when talking about error handling.
I'll `assert` for now... because yes, only 1 "alive" object should be found at a given moment.